onmt/io/IO.py

```diff
@@ -234,29 +244,49 @@ def build_vocab(train_datasets, data_type, share_vocab,
     Returns:
         Dict of Fields
     """
-    # All datasets have same fields, get the first one is OK.
-    fields = train_datasets[0].fields
-
-    fields["tgt"].build_vocab(*train_datasets, max_size=tgt_vocab_size,
-                              min_freq=tgt_words_min_frequency)
-    for j in range(train_datasets[0].n_tgt_feats):
-        fields["tgt_feat_" + str(j)].build_vocab(*train_datasets)
+    print("Building vocab:")
+    counter = {}
+    for k in fields:
+        counter[k] = Counter()
+
+    for path in train_datasets:
+        data = torch.load(path)
+        for x in data.examples:
+            for k in fields:
+                ex = getattr(x, k)
+                if not fields[k].sequential:
+                    ex = [ex]
+                counter[k].update(ex)
 
+    # All datasets have same fields, get the first one is OK.
+    _build_field_vocab(fields["tgt"], counter["tgt"],
+                       max_size=tgt_vocab_size,
+                       min_freq=tgt_words_min_frequency)
+    print("Building tgt size", len(fields["tgt"].vocab))
+    for j in range(data.n_tgt_feats):
+        key = "tgt_feat_" + str(j)
+        _build_field_vocab(fields[key], counter[key])
+        print("Building", key, "size", len(fields[key].vocab))
     if data_type == 'text':
-        fields["src"].build_vocab(*train_datasets, max_size=src_vocab_size,
-                                  min_freq=src_words_min_frequency)
-        for j in range(train_datasets[0].n_src_feats):
-            fields["src_feat_" + str(j)].build_vocab(*train_datasets)
+        _build_field_vocab(fields["src"], counter["src"],
+                           max_size=src_vocab_size,
+                           min_freq=src_words_min_frequency)
+        print("Building src size", len(fields["src"].vocab))
+        for j in range(data.n_src_feats):
+            key = "src_feat_" + str(j)
+            _build_field_vocab(fields[key], counter[key])
+            print("Building", key, "size", len(fields[key].vocab))
 
         # Merge the input and output vocabularies.
         if share_vocab:
             # `tgt_vocab_size` is ignored when sharing vocabularies
+
             merged_vocab = merge_vocabs(
                 [fields["src"].vocab, fields["tgt"].vocab],
                 vocab_size=src_vocab_size)
             fields["src"].vocab = merged_vocab
             fields["tgt"].vocab = merged_vocab
-
+            print("Merging src and tgt")
```

Comment:
This method is cool. And this is where I suspect why it fails to save memory. 

However, during the preprocess sharding work, I've tried to pass dataset generator( lazy `torch.load()`) into `preprocess.py::build_save_vocab()` (ie. passing pt list instead of dataset list),  but the result is weird - it doesn't save memory but uses more! 

But it is worth doing so, especially we now have materialized `dataset.examples` into a list in the `torchtext work`, keeping them inside memory is against our goal.

Reply:
I checked. It saves a lot of memory. There must have been a leak somewhere before.